kratos/containers/container_expression/expressions/expression_iterator.h

```diff
@@ -0,0 +1,283 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Suneth Warnakulasuriya
+//
+
+#pragma once
+
+// System includes
+
+// Project includes
+#include "includes/define.h"
+#include "containers/container_expression/expressions/expression.h"
+#include "containers/container_expression/expressions/literal/literal_flat_expression.h"
+
+namespace Kratos {
+
+///@name Kratos Classes
+///@{
+
+/**
+ * @brief Iterator class for expressions generic lazy expressions
+ *
+ */
+template<class TExpressionType>
+class ExpressionIterator {
+public:
+    ///@name Type definitions
+    ///@{
+
+    using IndexType = std::size_t;
+
+    using DataType = double;
+
+    using ConstIteratorType = ExpressionIterator<Expression>;
+
+    KRATOS_CLASS_POINTER_DEFINITION(ExpressionIterator);
+
+    ///@}
+    ///@name Life cycle
+    ///@{
+
+    ExpressionIterator(Expression::Pointer pExpression)
+        : mpExpression(pExpression),
+          mEntityIndex(0),
+          mEntityDataBeginIndex(0),
+          mComponentIndex(0),
+          mFlattenedShapeSize(pExpression->GetFlattenedShapeSize())
```

Comment:
flattened shape size? Isn't that **always** 1 by definition?

I'd stick to something like `mFlatSize`.

Reply:
changed to `mItemComponentCount`